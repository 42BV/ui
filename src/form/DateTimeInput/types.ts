// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
