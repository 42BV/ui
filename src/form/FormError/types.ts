export type OnChange = (hasErrors: boolean) => void;