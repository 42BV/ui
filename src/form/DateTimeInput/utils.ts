import moment from 'moment';
import { InputMask } from '../Input/Input';
import { DateFormat, TimeFormat } from './types';

export function combineFormat(
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): string {
  if (dateFormat !== false && timeFormat !== false) {
    return `${dateFormat} ${timeFormat}`;
  } else if (dateFormat !== false) {
    return dateFormat;
  } else if (timeFormat !== false) {
    return timeFormat;
  }

  throw new Error(
    'DateTimeInput: dateFormat and timeFormat cannot both be false'
  );
}

export function formatToMask(
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): InputMask {
  const dateMask = dateFormatToMask(dateFormat);

  const timeMask = timeFormatToMask(timeFormat);

  const dateMaskIsDefined = dateMask.length > 0;

  if (dateMaskIsDefined && timeMask.length > 0) {
    return [...dateMask, ' ', ...timeMask];
  } else if (dateMaskIsDefined) {
    return dateMask;
  } else {
    return timeMask;
  }
}

function timeFormatToMask(timeFormat: TimeFormat): InputMask {
  if (timeFormat === false) {
    return [];
  }

  return timeFormat.split('').map((char) => (char === ':' ? ':' : /\d/));
}

function dateFormatToMask(dateFormat: DateFormat): InputMask {
  if (dateFormat === false) {
    return [];
  }

  const separator = extractSeparator(dateFormat);
  return dateFormat
    .split('')
    .map((char) => (char === separator ? separator : /\d/));
}

function extractSeparator(dateFormat: string): string {
  if (dateFormat.includes('-')) {
    return '-';
  }

  if (dateFormat.includes('/')) {
    return '/';
  }

  if (dateFormat.includes('.')) {
    return '.';
  }

  throw new Error('DateTimeInput: cannot extract separator from dateFormat');
}

export function isDate(
  value: string,
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): boolean {
  // _ indicates a masks
  if (value.includes('_')) {
    return false;
  } else {
    const date = moment(
      value.trim(), // value includes an empty char at the back for some reason.
      combineFormat(dateFormat, timeFormat)
    );

    return date.isValid();
  }
}
