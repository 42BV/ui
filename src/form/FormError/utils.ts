import { ValidationError } from '@42.nl/jarb-final-form';
import { MetaError } from '../types';
import {
  Translation,
  getTranslator
} from '../../utilities/translation/translator';

export function keyForError(error: MetaError) {
  if (typeof error === 'string') {
    return error;
  }

  if ((error as ValidationError).type === undefined) {
    return (error as Translation).key;
  }

  return (error as ValidationError).type;
}

export function errorMessage(error: MetaError): string {
  const translator = getTranslator();

  if (typeof error === 'string') {
    return translator({ key: error, fallback: error });
  }

  if ((error as ValidationError).type === undefined) {
    return translator(error as Translation);
  }

  const validationError = error as ValidationError;

  switch (validationError.type) {
    case 'ERROR_REQUIRED':
      return translator({
        key: 'JarbFinalForm.VALIDATION.REQUIRED',
        data: validationError,
        fallback: `${validationError.label} is required`
      });
    case 'ERROR_MINIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MINIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be bigger than ${validationError.reasons.minimumLength} characters`
      });
    case 'ERROR_MAXIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAXIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be smaller than ${validationError.reasons.maximumLength} characters`
      });
    case 'ERROR_MIN_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MIN_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be more than ${validationError.reasons.minValue}`
      });
    case 'ERROR_MAX_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAX_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be less than ${validationError.reasons.maxValue}`
      });
    case 'ERROR_NUMBER':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER',
        data: validationError,
        fallback: `${validationError.label} is not a number`
      });
    case 'ERROR_NUMBER_FRACTION':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER_FRACTION',
        data: validationError,
        fallback: `${validationError.label} is not a number. Number may have ${validationError.reasons.fractionLength} digits behind the comma`
      });

    default:
      return translator({
        key: 'FormError.UNKNOWN_ERROR',
        fallback: 'Something is wrong'
      });
  }
}
