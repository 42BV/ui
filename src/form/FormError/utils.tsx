import React from 'react';
import { ValidationError } from '@42.nl/jarb-final-form';
import { MetaError } from '../types';
import {
  getTranslator,
  Translation
} from '../../utilities/translation/translator';

export function errorMessage(error: MetaError): React.ReactNode {
  const translator = getTranslator();

  // We consider it translated already
  if (typeof error === 'string' || React.isValidElement(error)) {
    return error;
  }

  if ((error as ValidationError).type === undefined) {
    return translator(error as Translation);
  }

  const validationError = error as ValidationError;

  switch (validationError.type) {
    case 'ERROR_REQUIRED':
      return translator({
        key: 'JarbFinalForm.VALIDATION.REQUIRED',
        data: validationError,
        fallback: `${validationError.label} is required`
      });
    case 'ERROR_MINIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MINIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be bigger than ${validationError.reasons.minimumLength} characters`
      });
    case 'ERROR_MAXIMUM_LENGTH':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAXIMUM_LENGTH',
        data: validationError,
        fallback: `${validationError.label} must be smaller than ${validationError.reasons.maximumLength} characters`
      });
    case 'ERROR_MIN_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MIN_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be more than ${validationError.reasons.minValue}`
      });
    case 'ERROR_MAX_VALUE':
      return translator({
        key: 'JarbFinalForm.VALIDATION.MAX_VALUE',
        data: validationError,
        fallback: `${validationError.label} must be less than ${validationError.reasons.maxValue}`
      });
    case 'ERROR_NUMBER':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER',
        data: validationError,
        fallback: `${validationError.label} is not a number`
      });
    case 'ERROR_NUMBER_FRACTION':
      return translator({
        key: 'JarbFinalForm.VALIDATION.NUMBER_FRACTION',
        data: validationError,
        fallback: `${validationError.label} is not a number. Number may have ${validationError.reasons.fractionLength} digits behind the comma`
      });

    default:
      return translator({
        key: 'FormError.UNKNOWN_ERROR',
        fallback: 'Something is wrong'
      });
  }
}
