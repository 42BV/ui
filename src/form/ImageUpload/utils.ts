import picaFn from 'pica';
import { ImageUploadCrop } from './ImageUpload';
import { Buffer } from 'buffer';
Buffer.from('anything','base64');

let picaInstance: any;

export function getPicaInstance() {
  if (picaInstance) {
    return picaInstance;
  }

  picaInstance = picaFn();

  return picaInstance;
}

export function dataUrlToFile(dataUrl: string, fileName: string): File {
  const byteString = Buffer.from(dataUrl.split(',')[1], 'base64');

  // separate out the mime component
  const mimeString = dataUrl.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to a typed array
  const bytes = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    bytes[i] = byteString[i];
  }

  return new File([ bytes ], fileName, { type: mimeString });
}

type AvatarEditorConfig = {
  borderRadius: number;
  width: number;
  height: number;
};

export function cropToAvatarEditorConfig(
  crop: ImageUploadCrop
): AvatarEditorConfig {
  if (crop.type === 'rect') {
    const { width, height } = crop;

    return { borderRadius: 0, width, height };
  } else {
    const { size } = crop;

    /* 
      When Crop is 'circle' a ridiculously high border radius is needed 
      so crop.size of different sizes show a perfect circle. Not sure
      why this occurs.
    */
    const borderRadius = 9000000;

    return { borderRadius, width: size, height: size };
  }
}

// Calculate the next 'scale' to use based on how hard the user scrolled
export function calculateScale(scale: number, delta: number): number {
  const zoomFactor = Math.max(Math.abs(delta) / 1000, 0.1);
  const nextScale = delta > 0 ? scale - zoomFactor : scale + zoomFactor;
  return Math.min(10, Math.max(1, nextScale));
}

// Replace any part after the last dot in the filename with .png
export function replaceFileExtension(fileName: string): string {
  return fileName.replace(/\.\w+$/, '.png');
}

export function readFile(file: File, callback: (result: string) => void) {
  const reader = new FileReader();
  reader.onloadend = () => {
    /* istanbul ignore else */
    if (typeof reader.result === 'string') {
      callback(reader.result);
    }
  };
  reader.readAsDataURL(file);
}
