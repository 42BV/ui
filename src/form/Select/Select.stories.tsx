import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Select, { JarbSelect } from './Select';
import { FinalForm, Form, resolveAfter } from '../story-utils';
import { pageOfUsers, userUser, randomUser } from '../../test/fixtures';
import { User } from '../../test/types';

interface SubjectOption {
  value: string;
  label: string;
}

storiesOf('Form|Select', module)
  .add('defined options', () => {
    return (
      <Form>
        <Select
          id="subject"
          label="Subject"
          placeholder="Please enter your subject"
          optionForValue={(option: SubjectOption) => option.label}
          isOptionEnabled={option => option.value !== 'awesome'}
          options={[
            { value: 'awesome', label: 'Awesome shit' },
            { value: 'super', label: 'Super shit' },
            { value: 'great', label: 'Great shit' },
            { value: 'good', label: 'good shit' }
          ]}
          onChange={value => action(`You entered ${value}`)}
        />
      </Form>
    );
  })
  .add('fetched options', () => {
    const [subject, setSubject] = useState<User>(randomUser());

    return (
      <Form>
        <Select
          id="subject"
          label="Subject"
          placeholder="Please enter your subject"
          optionForValue={(user: User) => user.email}
          options={() => resolveAfter(pageOfUsers)}
          value={subject}
          onChange={setSubject}
        />
      </Form>
    );
  })
  .add('custom isOptionEqual', () => {
    return (
      <Form>
        <Select
          id="subject"
          label="Subject"
          placeholder="Please enter your subject"
          value={userUser}
          optionForValue={(user: User) => user.email}
          options={() => Promise.resolve(pageOfUsers)}
          isOptionEqual={(a: User, b: User) => a.id === b.id}
          onChange={value => action(`You entered ${value}`)}
        />
      </Form>
    );
  })
  .add('jarb', () => {
    return (
      <FinalForm>
        <JarbSelect
          id="subject"
          name="subject"
          label="Subject"
          placeholder="Please enter your subject"
          optionForValue={(option: SubjectOption) => option.label}
          isOptionEnabled={option => option.value !== 'awesome'}
          options={[
            { value: 'awesome', label: 'Awesome shit' },
            { value: 'super', label: 'Super shit' },
            { value: 'great', label: 'Great shit' },
            { value: 'good', label: 'good shit' }
          ]}
          jarb={{
            validator: 'Hero.name',
            label: 'Subject'
          }}
        />
      </FinalForm>
    );
  })
  .add('default option', () => {
    const [value, setValue] = useState({ value: 'great', label: 'Great shit' });

    return (
      <Form>
        <Select<SubjectOption>
          id="subject"
          label="Subject"
          value={value}
          placeholder="Please enter your subject"
          optionForValue={(option: SubjectOption) => option.label}
          options={[
            { value: 'awesome', label: 'Awesome shit' },
            { value: 'super', label: 'Super shit' },
            { value: 'great', label: 'Great shit' },
            { value: 'good', label: 'good shit' }
          ]}
          onChange={setValue}
        />
      </Form>
    );
  });
