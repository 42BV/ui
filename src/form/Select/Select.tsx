import React, { Component } from 'react';
import { FormGroup, Label, Input as RSInput } from 'reactstrap';
import { get, constant, isArray } from 'lodash';
import { InputType } from 'reactstrap/lib/Input';
import { Page } from '@42.nl/spring-connect';

import withJarb from '../withJarb/withJarb';
import Spinner from '../../core/Spinner/Spinner';
import {
  Color,
  OptionForValue,
  OptionEnabledCallback,
  OptionsFetcher
} from '../types';
import { t } from '../../utilities/translation/translation';

export interface Text {
  /**
   * The message to show when the select is loading. Defaults
   * to `loading...`
   */
  loadingMessage?: string;
}
interface Props<T> {
  /**
   * The id of the form element.
   */
  id: string;

  /**
   * The value that the form element currently has.
   */
  value?: T;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: T) => void;

  /**
   * Optional callback for when the form element is blurred.
   */
  onBlur?: () => void;

  /**
   * Is either an array of options or a callback which fetches
   * the options asynchronously.
   */
  options: OptionsFetcher<T> | T[];

  /**
   * Callback to convert an value of type T to an option to show
   * to the user.
   */
  optionForValue: OptionForValue<T>;

  /**
   * Optional callback which is called for every option to determine
   * if the option can be selected. By default all options can be
   * selected.
   */
  isOptionEnabled?: OptionEnabledCallback<T>;

  /**
   * Optionally the error message to render.
   */
  error?: React.ReactNode;

  /**
   * Optionally the color of the FormGroup.
   */
  color?: Color;

  /**
   * Whether or not the form element is currently valid.
   */
  valid?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * The label of the form element.
   */
  label: string;

  /**
   * The placeholder of the form element.
   */
  placeholder: string;

  /**
   * Optionally customized text within the component.
   * This text should already be translated.
   */
  text?: Text;
}

interface State<T> {
  options: T[];
  loading: boolean;
}

/**
 * Select is a form element for which the value can be selected
 * from a limited range.
 */
export default class Select<T> extends Component<Props<T>, State<T>> {
  constructor(props: Props<T>) {
    super(props);

    const { options } = props;

    if (isArray(options)) {
      this.state = {
        options,
        loading: false
      };
    } else {
      this.state = {
        options: [],
        loading: true
      };
    }
  }

  async componentDidMount() {
    const options = this.props.options;

    if (isArray(options) === false) {
      const fetcher = options as OptionsFetcher<T>;

      const page: Page<T> = await fetcher();
      this.setState({ options: page.content, loading: false });

      this.setInitialValueAfterloadingOptions(page);
    }
  }

  setInitialValueAfterloadingOptions(page: Page<T>) {
    const { value, optionForValue } = this.props;

    // If there is a value try to see if it is still in the content
    if (value) {
      const label = optionForValue(value);

      if (!page.content.some(item => optionForValue(item) === label)) {
        // The value no longer appears in the options, so we select
        // the first option by default, overriding any value it
        // had already.
        this.props.onChange(page.content[0]);
      }
    } else {
      // If there is no value, select the first option by default
      this.props.onChange(page.content[0]);
    }
  }

  selectDefaultOption(option?: HTMLOptionElement | null) {
    // select the default option when no other option is chosen.
    if (option && this.props.value === undefined) {
      option.selected = true;
    }
  }

  render() {
    const { id, error, color, label, text = {}, className = '' } = this.props;
    const { loading } = this.state;

    return (
      <FormGroup className={className} color={color}>
        <Label for={id}>{label}</Label>
        {loading ? (
          <div>
            <Spinner color="black" size={16} />
            <span>
              {t({
                key: 'Select.LOADING',
                fallback: 'Loading...',
                overrideText: text.loadingMessage
              })}
            </span>
          </div>
        ) : (
          this.renderInput()
        )}

        {error}
      </FormGroup>
    );
  }

  renderInput() {
    const {
      id,
      placeholder,
      value,
      valid,
      onBlur,
      onChange,
      optionForValue
    } = this.props;

    const { options } = this.state;

    const isOptionEnabled = get(this.props, 'isOptionEnabled', constant(true));
    const inputProps = {
      id,
      valid,
      invalid: valid === false ? true : undefined,
      type: 'select' as InputType,
      placeholder,
      onChange: (event: { target: { value: string } }) => {
        const index = parseInt(event.target.value, 10);
        onChange(options[index]);
      },
      onBlur,
      className: value === undefined ? 'showing-placeholder' : ''
    };

    // @ts-ignore
    const indexOfValue = options.findIndex(option => option === value);

    return (
      <RSInput value={indexOfValue} {...inputProps}>
        <option disabled ref={option => this.selectDefaultOption(option)}>
          {placeholder}
        </option>

        {options.map((option, index) => {
          const label = optionForValue(option);

          return (
            <option
              key={label}
              // @ts-ignore
              value={index}
              disabled={!isOptionEnabled(option)}
            >
              {label}
            </option>
          );
        })}
      </RSInput>
    );
  }
}
/**
 * Variant of the Select which can be used in a Jarb context.
 */
export const JarbSelect = withJarb<string, string, Props<any>>(Select);
