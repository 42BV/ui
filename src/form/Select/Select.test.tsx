import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import Select, { Text } from './Select';
import { User } from '../../test/types';
import {
  adminUser,
  coordinatorUser,
  userUser,
  pageOfUsers
} from '../../test/fixtures';
import { OptionEnabledCallback } from '../option';
import { pageWithContent } from '../../test/utils';

describe('Component: Select', () => {
  let select: ShallowWrapper;

  let onChangeSpy: jest.Mock<any, any>;
  let onBlurSpy: jest.Mock<any, any>;

  function setup({
    value,
    isOptionEnabled,
    text
  }: {
    value?: User;
    isOptionEnabled?: OptionEnabledCallback<User>;
    text?: Text;
  }) {
    onChangeSpy = jest.fn();
    onBlurSpy = jest.fn();

    select = shallow(
      <Select
        id="subject"
        label="Subject"
        placeholder="Please enter your subject"
        text={text}
        isOptionEnabled={isOptionEnabled}
        optionForValue={(user: User) => user?.email}
        options={[adminUser, coordinatorUser, userUser]}
        value={value}
        onChange={onChangeSpy}
        onBlur={onBlurSpy}
        error="Some error"
        valid={false}
      />
    );
  }

  describe('ui', () => {
    test('with value', () => {
      setup({ value: adminUser, isOptionEnabled: undefined });

      expect(toJson(select)).toMatchSnapshot(
        'Component: Select => ui => with value'
      );
    });

    describe('loading', () => {
      test('with custom text', () => {
        setup({ value: adminUser, text: { loadingMessage: 'Custom loading' } });

        // @ts-ignore
        select.setState({ loading: true });

        expect(select.find('Spinner').exists()).toBe(true);
        expect(select.find('span').text()).toBe('Custom loading');
      });

      test('with default text', () => {
        setup({ value: adminUser });

        // @ts-ignore
        select.setState({ loading: true });

        expect(select.find('Spinner').exists()).toBe(true);
        expect(select.find('span').text()).toBe('Loading...');
      });
    });

    test('empty value string, should show placeholder', () => {
      setup({ value: undefined, isOptionEnabled: undefined });

      const rsInput = select.find('Input');

      expect(rsInput.props().className).toBe('showing-placeholder');
    });

    test('when value is not in options select nothing', () => {
      setup({
        value: {
          id: -1,
          email: 'none',
          active: false,
          roles: []
        },
        isOptionEnabled: undefined
      });

      const rsInput = select.find('Input');

      expect(rsInput.props().value).toBe(undefined);
    });
  });

  describe('constructor', () => {
    test('when options is an array use that options and set loading to false', () => {
      // @ts-ignore
      const select = new Select({ options: [adminUser] });

      expect(select.state).toEqual({ loading: false, options: [adminUser] });
    });

    test('when options is a function set options to empty and loading to true', () => {
      // @ts-ignore
      const select = new Select({ options: jest.fn() });

      expect(select.state).toEqual({ loading: true, options: [] });
    });
  });

  describe('componentDidMount', () => {
    test('when options is an array do nothing', async done => {
      const onChange = jest.fn();

      // @ts-ignore
      const select = new Select({
        options: [adminUser, coordinatorUser, userUser],
        onChange
      });

      jest.spyOn(select, 'setState').mockImplementation(() => undefined);

      try {
        await select.componentDidMount();

        expect(select.setState).toHaveBeenCalledTimes(0);

        expect(onChange).toHaveBeenCalledTimes(0);

        done();
      } catch (e) {
        console.error(e);
        done.fail();
      }
    });

    describe('when options is a function', () => {
      test('when options is a function set options to empty and loading to true', async done => {
        const options = () => Promise.resolve(pageOfUsers);

        // @ts-ignore
        const select = new Select({ options, onChange: jest.fn() });

        jest.spyOn(select, 'setState').mockImplementation(() => undefined);

        try {
          await select.componentDidMount();

          expect(select.setState).toHaveBeenCalledTimes(1);
          expect(select.setState).toHaveBeenCalledWith({
            loading: false,
            options: [adminUser, coordinatorUser, userUser]
          });

          done();
        } catch (e) {
          console.error(e);
          done.fail();
        }
      });

      describe('setting of initial value after loading options', () => {
        test('do nothing when value is set and can be found in the loaded options', async done => {
          const onChange = jest.fn();
          const options = () => Promise.resolve(pageOfUsers);

          // @ts-ignore
          const select = new Select({ options, onChange });
          select.props.value = adminUser;
          select.props.optionForValue = (user: User) => user?.email;

          jest.spyOn(select, 'setState').mockImplementation(() => undefined);

          try {
            await select.componentDidMount();

            expect(onChange).toHaveBeenCalledTimes(0);

            done();
          } catch (e) {
            console.error(e);
            done.fail();
          }
        });

        test('select first option when value is set but not in the loaded options', async done => {
          const onChange = jest.fn();
          const options = () =>
            Promise.resolve(pageWithContent([adminUser, coordinatorUser]));

          // @ts-ignore
          const select = new Select({ options, onChange });
          select.props.value = userUser;
          select.props.optionForValue = (user: User) => user?.email;

          jest.spyOn(select, 'setState').mockImplementation(() => undefined);

          try {
            await select.componentDidMount();

            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange).toBeCalledWith(adminUser);

            done();
          } catch (e) {
            console.error(e);
            done.fail();
          }
        });

        test('select first option when value is not set', async done => {
          const onChange = jest.fn();
          const options = () => Promise.resolve(pageOfUsers);

          // @ts-ignore
          const select = new Select({ options, onChange });

          jest.spyOn(select, 'setState').mockImplementation(() => undefined);

          try {
            await select.componentDidMount();

            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange).toBeCalledWith(adminUser);

            done();
          } catch (e) {
            console.error(e);
            done.fail();
          }
        });
      });
    });
  });

  describe('selectDefaultOption', () => {
    it('should select the placeholder as the default value when value is empty string', () => {
      setup({ value: undefined, isOptionEnabled: undefined });

      const defaultOption = select.find('option').first();
      expect(defaultOption.text()).toBe('Please enter your subject');

      const option = { selected: false };
      // @ts-ignore
      defaultOption.getElement().ref(option);

      expect(option.selected).toBe(true);

      // @ts-ignore
      expect(option.value).toBe(undefined);
    });

    it('should not select the placeholder as the default value when value is not an empty string', () => {
      setup({ value: adminUser, isOptionEnabled: undefined });

      const defaultOption = select.find('option').first();
      expect(defaultOption.text()).toBe('Please enter your subject');

      const option = { selected: false };
      // @ts-ignore
      defaultOption.getElement().ref(option);

      expect(option.selected).toBe(false);
    });

    it('should not select the placeholder as the default value when the option is not defined', () => {
      setup({ value: undefined, isOptionEnabled: undefined });

      const defaultOption = select.find('option').first();
      expect(defaultOption.text()).toBe('Please enter your subject');

      const option = undefined;
      // @ts-ignore
      defaultOption.getElement().ref(option);
    });
  });

  describe('events', () => {
    test('onChange', () => {
      setup({ value: undefined, isOptionEnabled: undefined });

      const rsInput = select.find('Input');

      // @ts-ignore
      rsInput.props().onChange({ target: { value: 0 } });

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(adminUser);
    });

    test('onBlur', () => {
      setup({ value: undefined, isOptionEnabled: undefined });

      const rsInput = select.find('Input');
      // @ts-ignore
      rsInput.props().onBlur();

      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });

    describe('isOptionEnabled', () => {
      it('should when "isOptionEnabled" is not defined always be true', () => {
        setup({ value: adminUser, isOptionEnabled: undefined });

        const options = select.find('option');
        expect(options.length).toBe(4);

        // Other options should be enabled
        expect(options.at(1).props().disabled).toBe(false);
        expect(options.at(2).props().disabled).toBe(false);
        expect(options.at(3).props().disabled).toBe(false);
      });

      it('should when "isOptionEnabled" is defined use that to determine if the option is enabled', () => {
        const isOptionEnabledSpy = jest.fn();

        // Disabled all option now
        isOptionEnabledSpy.mockReturnValue(false);

        setup({ value: undefined, isOptionEnabled: isOptionEnabledSpy });

        const options = select.find('option');
        expect(options.length).toBe(4);

        // Other options should be enabled
        expect(options.at(1).props().disabled).toBe(true);
        expect(options.at(2).props().disabled).toBe(true);
        expect(options.at(3).props().disabled).toBe(true);

        expect(isOptionEnabledSpy).toHaveBeenCalledTimes(3);
        expect(isOptionEnabledSpy.mock.calls).toEqual([
          [adminUser],
          [coordinatorUser],
          [userUser]
        ]);
      });
    });
  });

  describe('value changes', () => {
    test('becomes empty', () => {
      setup({ value: userUser, isOptionEnabled: undefined });

      let rsInput = select.find('Input');
      expect(rsInput.props().value).toBe(2);

      select.setProps({ value: undefined });

      rsInput = select.find('Input');
      expect(rsInput.props().value).toBe(undefined);
    });

    test('becomes filled', () => {
      setup({ value: undefined, isOptionEnabled: undefined });

      let rsInput = select.find('Input');
      expect(rsInput.props().value).toBe(undefined);

      select.setProps({ value: coordinatorUser });

      rsInput = select.find('Input');
      expect(rsInput.props().value).toBe(1);
    });
  });
});
