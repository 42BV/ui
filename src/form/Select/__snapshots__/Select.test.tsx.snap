// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Select ui with value: Component: Select => ui => with value 1`] = `
<FormGroup
  className=""
  tag="div"
>
  <Label
    for="subject"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    Subject
  </Label>
  <Input
    className=""
    id="subject"
    invalid={true}
    onBlur={[MockFunction]}
    onChange={[Function]}
    placeholder="Please enter your subject"
    type="select"
    valid={false}
    value={0}
  >
    <option
      disabled={true}
    >
      Please enter your subject
    </option>
    <option
      disabled={false}
      key="admin@42.nl"
      value={0}
    >
      admin@42.nl
    </option>
    <option
      disabled={false}
      key="coordinator@42.nl"
      value={1}
    >
      coordinator@42.nl
    </option>
    <option
      disabled={false}
      key="user@42.nl"
      value={2}
    >
      user@42.nl
    </option>
  </Input>
  Some error
</FormGroup>
`;
